base36.go

```diff
@@ -0,0 +1,158 @@
+/*
+
+Package base36 provides a reasonably fast implementation of a binary base36 codec.
+
+*/
+package base36
+
+// Simplified code based on https://godoc.org/github.com/mr-tron/base58
+// which in turn is based on https://github.com/trezor/trezor-crypto/commit/89a7d7797b806fac
+
+import (
+	"fmt"
+	"math"
+)
+
+const UcAlphabet = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+const LcAlphabet = "0123456789abcdefghijklmnopqrstuvwxyz"
+
+var encExpansion int
+var revAlphabet [256]uint64
+
+func init() {
+	for i := range revAlphabet {
+		revAlphabet[i] = 36
+	}
+	for i, c := range UcAlphabet {
+		revAlphabet[byte(c)] = uint64(i)
+		if c > '9' {
+			revAlphabet[byte(c)+32] = uint64(i)
+		}
+	}
+
+	encExpansion = int(
+		math.Ceil(
+			math.Log(256) / math.Log(36),
+		),
+	)
+}
+
+// EncodeToStringUc encodes the given byte-buffer as base36 using [0-9A-Z] as
+// the digit-alphabet
+func EncodeToStringUc(b []byte) string { return encode(b, UcAlphabet) }
+
+// EncodeToStringLc encodes the given byte-buffer as base36 using [0-9a-z] as
+// the digit-alphabet
+func EncodeToStringLc(b []byte) string { return encode(b, LcAlphabet) }
+
+func encode(inBuf []byte, al string) string {
+
+	// As a polar opposite to the base58 implementation, using a uint32 here is
+	// significantly slower
+	var carry uint
+
+	var encIdx, valIdx, zcnt, high int
+
+	inSize := len(inBuf)
+	for zcnt < inSize && inBuf[zcnt] == 0 {
+		zcnt++
+	}
+
+	encSize := (inSize - zcnt) * encExpansion
+
+	// Allocate one big buffer up front
+	// Note: pools *DO NOT* help, the overhead of zeroing the val-half (see below)
+	// kills any performance gain to be had
+	outBuf := make([]byte, (zcnt + encSize*2))
```

Comment:
This should end up getting allocated on the stack, anyways, right?

Reply:
Yeah, since we `string()` it in the end it should not escape. But gives me an idea to tripple check the `-m -m` output